### ISSUE TITLE
Numpy version compatibility

### ISSUE THREAD

**Issue by David-Baddeley** [python-microscopy/python-microscopy#74]
**[Original report](https://bitbucket.org/david_baddeley/python-microscopy/issue/74) by David Baddeley (Bitbucket: [david_baddeley](https://bitbucket.org/david_baddeley), ).**

----------------------------------------

We're currently tied to numpy 1.11 for compatibility with mayavi, but also because there are several places where we index with non-integers and rely on numpy to cast (which is deprecated in more recent versions).

Updating the dependency to a more recent version is desirable in the mid-to long term, and indeed most things run if numpy is updated. I'll suggest a new target of 1.14. I've been running 1.14 for a while and slowly fixing things as I come across them.

This issue will list things which are known to be broken in terms of numpy compatibility.

numpy 1.11
-----------
- Zach's triangle code is broken (due to new np.unique arguments)

numpy 1.14
------------
- optical flow visualization is broken
- mayavi package dependencies (although mayavi itself seems to work)
- pyme-depends package
- others ??

**Comment by David-Baddeley** [comment]
**Original comment by Andrew Barentine (Bitbucket: [AndrewBarentine](https://bitbucket.org/AndrewBarentine), ).**

----------------------------------------

@david_baddeley does the pyme-depends package actually broken on 1.14? Would be nice to have pymecompress installable via conda for people working with numpy beyond 1.11

**Comment by David-Baddeley** [comment]
**Original comment by David Baddeley (Bitbucket: [david_baddeley](https://bitbucket.org/david_baddeley), ).**

----------------------------------------

Yep - pyme-depends has numpy=1.11 as a dependency, although you can install numpy=1.14 over the top of it once installed. This is probably the right behaviour until we are a little more certain that everything else works with numpy=1.14 (i.e. numpy=1.11 should still be the default, and anything else at your own risk and needs deliberate action). Compiling pyme-depends against 1.14 would make this the default.

pymecompress is already compiled for numpy 1.14 on some platforms (OSX as a definite, possibly also windows, probably not linux at this point)

**Comment by David-Baddeley** [comment]
**Original comment by David Baddeley (Bitbucket: [david_baddeley](https://bitbucket.org/david_baddeley), ).**

----------------------------------------

New builds of pyme-compress will build against python 2.7, python 3.6, numpy 1.11 and numpy 1.14

**Comment by David-Baddeley** [comment]
**Original comment by Andrew Barentine (Bitbucket: [AndrewBarentine](https://bitbucket.org/AndrewBarentine), ).**

----------------------------------------

@david_baddeley , am I missing something or is there no clean way to install numpy 1.14 over the top of numpy 1.11 without it uninstalling pymecompress? Edit: Note that on python 2.7 there is no problem updating (removing pymecompress) and then installing pymecompress via distutils (though it says the version is 0.0.0).


```
#!python

$ dh5view something.h5

No handlers could be found for logger "/home/user/code/python-microscopy/PYME/config.pyc"

Gtk-Message: 15:24:31.350: Failed to load module "canberra-gtk-module"

Starting main loop

Loading data

filename == something.h5

something.h5r

RuntimeError: module compiled against API version 0xc but this version of numpy is 0xa

Traceback (most recent call last):

  File "/home/user/anaconda2/lib/python2.7/site-packages/wx-3.0-gtk2/wx/_core.py", line 16763, in <lambda>

    lambda event: event.callable(*event.args, **event.kw) )

  File "/home/user/code/python-microscopy/PYME/DSView/dsviewer.py", line 433, in LoadData

    vframe = DSViewFrame(im, None, im.filename, mode = mode)

  File "/home/user/code/python-microscopy/PYME/DSView/dsviewer.py", line 161, in __init__

    modules.loadMode(self.mode, self)

  File "/home/user/code/python-microscopy/PYME/DSView/modules/__init__.py", line 94, in loadMode

    loadModule(m, dsviewer)

  File "/home/user/code/python-microscopy/PYME/DSView/modules/__init__.py", line 69, in loadModule

    mod = __import__('.'.join(ml) + '.' + modName, fromlist=ml)

  File "/home/user/code/python-microscopy/PYME/DSView/modules/arrayView.py", line 23, in <module>

    from PYME.DSView.arrayViewPanel import *

  File "/home/user/code/python-microscopy/PYME/DSView/arrayViewPanel.py", line 36, in <module>

    from PYME.DSView.LUT import applyLUT

  File "/home/user/code/python-microscopy/PYME/DSView/LUT/__init__.py", line 24, in <module>

    from .lut import *

ImportError: numpy.core.multiarray failed to import

^C

$ conda list | grep numpy

numpy                     1.11.3           py27h3dfced4_4  

numpy-base                1.15.1           py27h81de0dd_0  

numpydoc                  0.8.0                    py27_0  

```


```
#!python

$ conda install numpy=1.14

Solving environment: done

## Package Plan ##

  environment location: /home/user/anaconda2

  added / updated specs: 

    - numpy=1.14

The following packages will be downloaded:

    package                    |            build

    ---------------------------|-----------------

    matplotlib-2.2.2           |   py27hb69df0a_2         6.5 MB

    numpy-1.14.6               |   py27h3b04361_4          46 KB

    mkl-2018.0.3               |                1       198.7 MB

    pixman-0.38.0              |       h7b6447c_0         618 KB

    qt-5.6.3                   |       h8bf5577_3        45.7 MB

    sqlite-3.26.0              |       h7b6447c_0         1.9 MB

    certifi-2019.3.9           |           py27_0         155 KB

    ca-certificates-2019.1.23  |                0         126 KB

    numpy-base-1.14.6          |   py27h81de0dd_4         4.1 MB

    conda-4.6.14               |           py27_0         2.1 MB

    libpng-1.6.36              |       hbc83047_0         346 KB

    pyqt-5.6.0                 |   py27h22d08a2_6         5.3 MB

    mkl_fft-1.0.6              |   py27h7dd41cf_0         148 KB

    pillow-5.4.1               |   py27h34e0f95_0         618 KB

    ------------------------------------------------------------

                                           Total:       266.2 MB

The following packages will be REMOVED:

    pymecompress:    0.1.2-py27_1          david_baddeley

The following packages will be UPDATED:

    ca-certificates: 2018.03.07-0                         --> 2019.1.23-0          

    cairo:           1.12.18-6                            --> 1.14.12-h8948797_3   

    certifi:         2018.10.15-py27_0                    --> 2019.3.9-py27_0      

    conda:           4.5.11-py27_0                        --> 4.6.14-py27_0        

    fontconfig:      2.11.1-6                             --> 2.13.0-h9420a91_0    

    freetype:        2.5.5-2                              --> 2.9.1-h8a8886c_1     

    harfbuzz:        0.9.39-1                             --> 1.8.8-hffaf4a1_0     

    libpng:          1.6.34-hb9fc6fc_0                    --> 1.6.36-hbc83047_0    

    matplotlib:      1.5.1-np111py27_0                    --> 2.2.2-py27hb69df0a_2 

    mkl_fft:         1.0.4-py27h4414c95_1                 --> 1.0.6-py27h7dd41cf_0 

    numpy:           1.11.3-py27h3dfced4_4                --> 1.14.6-py27h3b04361_4

    pango:           1.39.0-0                             --> 1.42.4-h049681c_0    

    pillow:          4.2.1-py27_0                         --> 5.4.1-py27h34e0f95_0 

    pixman:          0.32.6-0                             --> 0.38.0-h7b6447c_0    

    pyqt:            4.11.4-py27_4                        --> 5.6.0-py27h22d08a2_6 

    qt:              4.8.7-2                              --> 5.6.3-h8bf5577_3     

    sqlite:          3.25.2-h7b6447c_0                    --> 3.26.0-h7b6447c_0    

The following packages will be DOWNGRADED:

    mkl:             2019.0-118                           --> 2018.0.3-1           

    numpy-base:      1.15.1-py27h81de0dd_0                --> 1.14.6-py27h81de0dd_4

Proceed ([y]/n)?
```

**Comment by David-Baddeley** [comment]
**Original comment by David Baddeley (Bitbucket: [david_baddeley](https://bitbucket.org/david_baddeley), ).**

----------------------------------------

Depending on the platform, Pymecompress may not be built for all versions
of numpy yet.

Looking at the package spec, however, it looks like you also have a problem
with the version of numpy-base.

**Comment by David-Baddeley** [comment]
**Original comment by David Baddeley (Bitbucket: [david_baddeley](https://bitbucket.org/david_baddeley), ).**

----------------------------------------

Solution is to build conda packages for pymecompress for numpy=1.14 on linux. For some reason the build process is currently broken (might just be the old linux install I'm using to generate packages). 

This is on the TODO list, but a temporary workaround (as noted) is to just install pymecompress from source with `python setup.py install` **OR** use numpy=1.11. (one of the reasons we are still not compiling the core packages against numpy=1.14).

**Comment by David-Baddeley** [comment]
**Original comment by Zach Connerty-Marin (Bitbucket: [zacsimile](https://bitbucket.org/zacsimile), GitHub: [zacsimile](https://github.com/zacsimile)).**

----------------------------------------

When running on Windows, the newest version of VisGUI requries numpy=1.14. I can’t open h5r files without upgrading to numpy=1.14. The culprit seems to be rcond=None on line 64 of PYME.LMVIs.layers.base.

**Comment by David-Baddeley** [comment]
**Original comment by David Baddeley (Bitbucket: [david_baddeley](https://bitbucket.org/david_baddeley), ).**

----------------------------------------

Should now be reverted \(was added to avoid lots of spammy warnings, but we might need to accept them for now\)

**Comment by David-Baddeley** [comment]
**Original comment by Zach Connerty-Marin (Bitbucket: [zacsimile](https://bitbucket.org/zacsimile), GitHub: [zacsimile](https://github.com/zacsimile)).**

----------------------------------------

Need to install numpy=1.14 to run VisGUI on Python 3 on a fresh Mac install. No action is needed, I’m strictly documenting in case we run into this again, as pyme-depends still installs 1.11.3 by default. Traceback w/ numpy=1.11.3 is

‌

Traceback \(most recent call last\):

File "/anaconda3/envs/pyme3/bin/visgui", line 11, in <module>

load\_entry\_point\('PYME', 'console\_scripts', 'VisGUI'\)\(\)

File "/anaconda3/envs/pyme3/lib/python3.6/site-packages/pkg\_resources/\_\_init\_\_.py", line 489, in load\_entry\_point

return get\_distribution\(dist\).load\_entry\_point\(group, name\)

File "/anaconda3/envs/pyme3/lib/python3.6/site-packages/pkg\_resources/\_\_init\_\_.py", line 2843, in load\_entry\_point

return ep.load\(\)

File "/anaconda3/envs/pyme3/lib/python3.6/site-packages/pkg\_resources/\_\_init\_\_.py", line 2434, in load

return self.resolve\(\)

File "/anaconda3/envs/pyme3/lib/python3.6/site-packages/pkg\_resources/\_\_init\_\_.py", line 2440, in resolve

module = \_\_import\_\_\(self.module\_name, fromlist=\['\_\_name\_\_'\], level=0\)

File "/Users/zachcm/Code/mesh-halfedge/PYME/LMVis/VisGUI.py", line 61, in <module>

from PYME.recipes import recipeGui

File "/Users/zachcm/Code/mesh-halfedge/PYME/recipes/recipeGui.py", line 21, in <module>

from PYME.recipes import modules

File "/Users/zachcm/Code/mesh-halfedge/PYME/recipes/modules.py", line 22, in <module>

from . import skfilters

File "/Users/zachcm/Code/mesh-halfedge/PYME/recipes/skfilters.py", line 14, in <module>

import skimage.filters as skf

File "/anaconda3/envs/pyme3/lib/python3.6/site-packages/skimage/\_\_init\_\_.py", line 127, in <module>

from .util.dtype import \(img\_as\_float32,

File "/anaconda3/envs/pyme3/lib/python3.6/site-packages/skimage/util/\_\_init\_\_.py", line 6, in <module>

from .apply\_parallel import apply\_parallel

File "/anaconda3/envs/pyme3/lib/python3.6/site-packages/skimage/util/apply\_parallel.py", line 8, in <module>

import dask.array as da

File "/anaconda3/envs/pyme3/lib/python3.6/site-packages/dask/array/\_\_init\_\_.py", line 5, in <module>

from .core import \(Array, block, concatenate, stack, from\_array, store,

File "/anaconda3/envs/pyme3/lib/python3.6/site-packages/dask/array/core.py", line 29, in <module>

from . import chunk

File "/anaconda3/envs/pyme3/lib/python3.6/site-packages/dask/array/chunk.py", line 73, in <module>

nancumprod = npcompat.nancumprod

AttributeError: module 'dask.array.numpy\_compat' has no attribute 'nancumprod'

**Comment by David-Baddeley** [comment]
**Original comment by David Baddeley (Bitbucket: [david_baddeley](https://bitbucket.org/david_baddeley), ).**

----------------------------------------

This will be a skimage version problem - did you first install numpy 1.14,  and skimage and then downgrade?  We might need to change install instructions to start with miniconda rather than anaconda.

**Comment by David-Baddeley** [comment]
**Original comment by Zach Connerty-Marin (Bitbucket: [zacsimile](https://bitbucket.org/zacsimile), GitHub: [zacsimile](https://github.com/zacsimile)).**

----------------------------------------

This was on a fresh install of anaconda3 on my Mac. I typed

```
conda create -n pyme3 python=3
conda config --add channels david_baddeley
conda install pyme-depends
```

Then I got the traceback above. Then I typed

```
conda install numpy=1.14
```

and everything worked.

**Comment by David-Baddeley** [comment]
**Original comment by Zach Connerty-Marin (Bitbucket: [zacsimile](https://bitbucket.org/zacsimile), GitHub: [zacsimile](https://github.com/zacsimile)).**

----------------------------------------

Nuke and paved my Anaconda install and ran into this error again. You’re right, it’s an skimage problem. pyme-depends installs skimage 0.14.2 by default on Anaconda2 but 0.15.0 by default on Anaconda3. The newer skimage requires numpy >= 1.12.

**Comment by David-Baddeley** [comment]
**Original comment by Andrew Barentine (Bitbucket: [AndrewBarentine](https://bitbucket.org/AndrewBarentine), ).**

----------------------------------------

I couldn’t open VisGUI on numpy 1.11 due to the skimage problem @{5b68503b01fca93c25044e5b} mentioned without downgrading scikits-image all the way to 0.12.3

**Comment by David-Baddeley** [comment]
**Original comment by David Baddeley (Bitbucket: [david_baddeley](https://bitbucket.org/david_baddeley), ).**

----------------------------------------

Also ran into this on a fresh windows install yesterday. Code is now modified such that the wrong version of skimage will be tolerated \(i.e. won’t cause a hard failure\).

In theory, conda’s dependency management should have sorted this out and downgraded skimage automatically \(failure to do this means that either conda or the skimage package is broken\).

Short term workaround is to pin skimage to 0.12.x

Seriously considering bumping the default numpy version to 1.14 though, as I think we’re at the point where the few remaining broken things might not matter that much.

**Comment by David-Baddeley** [comment]
Hopefully  fixed - now build against both 1.11 and 1.14 but have >= pinning and should allow higher numpy versions to be installed over the top if needed . skimage packages were broken, and have been removed from conda (with the consequence that the only conda packages for skimage now depend on numpy >=1.16.
